src/coord/src/coord.rs

```diff
@@ -5182,54 +5183,68 @@ mod timeline {
     }
 
-    impl TimestampOracle {
-        /// Create a new timeline, starting at the indicated time.
-        pub fn new(initially: mz_repr::Timestamp) -> Self {
+    impl<T: super::CoordTimestamp> TimestampOracle<T> {
+        /// Create a new timeline, starting at the indicated time. `next` generates
+        /// new timestamps when invoked. The timestamps have no requirements, and can
+        /// retreat from previous invocations.
+        pub fn new<F>(initially: T, next: F) -> Self
+        where
+            F: Fn() -> T + 'static,
+        {
             Self {
-                state: TimestampOracleState::Writing(initially),
+                state: TimestampOracleState::Writing(initially.clone()),
                 advance_to: Some(initially),
+                next: Box::new(next),
             }
         }
 
         /// Acquire a new timestamp for writing.
         ///
-        /// This timestamp will be strictly greater than all prior values of `self.read_ts()`,
-        /// and less than or equal to all subsequent values of `self.read_ts()`.
-        pub fn write_ts(&mut self) -> mz_repr::Timestamp {
-            match self.state {
-                TimestampOracleState::Writing(ts) => ts,
+        /// This timestamp will be strictly greater than all prior values of
+        /// `self.read_ts()`, and less than or equal to all subsequent values of
+        /// `self.read_ts()`.
+        pub fn write_ts(&mut self) -> T {
+            match &self.state {
+                TimestampOracleState::Writing(ts) => ts.clone(),
                 TimestampOracleState::Reading(ts) => {
-                    self.state = TimestampOracleState::Writing(ts + 1);
-                    ts + 1
+                    let mut next = (self.next)();
+                    if next.less_equal(&ts) {
+                        next = ts.step_forward();
+                    }
+                    self.state = TimestampOracleState::Writing(next.clone());
+                    next
                 }
             }
         }
         /// Acquire a new timestamp for reading.
         ///
         /// This timestamp will be greater or equal to all prior values of `self.write_ts()`,
         /// and strictly less than all subsequent values of `self.write_ts()`.
-        pub fn read_ts(&mut self) -> mz_repr::Timestamp {
-            match self.state {
-                TimestampOracleState::Reading(ts) => ts,
+        pub fn read_ts(&mut self) -> T {
+            match &self.state {
+                TimestampOracleState::Reading(ts) => ts.clone(),
                 TimestampOracleState::Writing(ts) => {
-                    self.state = TimestampOracleState::Reading(ts);
-                    self.advance_to = Some(ts + 1);
+                    // Avoid rust borrow complaint.
+                    let ts = ts.clone();
+                    self.state = TimestampOracleState::Reading(ts.clone());
+                    self.advance_to = Some(ts.step_forward());
```

Comment:
This line advertises as complete the current read time (setting the write cursor ahead by that much), but it's now a bit awkward to reason about when the write cursor will next move forward. For example, on the read-to-write transition, we do not advance `self.advance_to` up to the new timestamp, if it is not `ts.step_forward()`. That might be fine, because no one can tell until you do a read, but I think maybe we want add `self.advance_to = Some(next.clone())` to `write_ts`, wdyt?

Reply:
Agree!